### PR TITLE
Move tests from core/providers/cuda/test/* to test/providers/cuda/ and refactor CUDA UT

### PR DESCRIPTION
**Pull request by guoyuhong**
### Description

1. Add a new test lib `onnxruntime_providers_cuda_ut` which is similar to `onnxruntime_providers_cuda` but `onnxruntime_providers_cuda_ut` is only built if `onnxruntime_BUILD_UNIT_TESTS` is set. We can call all CUDA UTs through this ut lib without affecting production lib `onnxruntime_providers_cuda`.
2. Move all test cases from `core/providers/cuda/test/` to `test/providers/cuda/`. These test cases are built into lib `onnxruntime_providers_cuda_ut` and run by `./onnxruntime_test_all --gtest_filter="*CUDA_EP_Unittest*"`. Since the lib is only for test, we can use gtest macros in the test cases. Previous implementation do not support using gtest lib in the CUDA UT cases.
3. The cmake code in `cmake/onnxruntime_providers.cmake` is refactored a bit. A new function `onnxruntime_add_object_library` is to build a object target. The 2 libs `onnxruntime_providers_cuda_ut` & `onnxruntime_providers_cuda` share most of the code, so the object files can be used in both libs, which helps reduce build time. Another function `config_cuda_provider_shared_module` is used to configure all 3 similar targets(onnxruntime_providers_cuda_obj/onnxruntime_providers_cuda/onnxruntime_providers_cuda_ut).
4. Refactored the test to call `testing::InitGoogleTest` & `RUN_ALL_TESTS` in `libonnxruntime_providers_cuda_ut.so`'s `TestAll`. After this change, we can see all the cases running in `CUDA_EP_Unittest.All`:
image

### Motivation and Context

After https://github.com/microsoft/onnxruntime/pull/13016, there are still test files in test/providers/cuda/ that are not moved to core/providers/cuda/test/ and the test cases are disabled. This PR helps to clean the unfinished TODOs.

Even through onnxruntime_shared_lib_test covers some test for CUDA provider. onnxruntime_shared_lib_test works like a coarse grain end-to-end test for CUDA provider. If CUDA unittest can run cases for a single component, this wound be helpful for CUDA developers.

### CODE REVIEW

**Comment by guoyuhong** [comment]
@souptc @wschin Would you please to review this PR?

**Comment by pranavsharma** [comment]
What's the motivation behind this change?

**Review by snnn — DISMISSED** [review]

**Comment by guoyuhong** [comment]
@pranavsharma 

> What's the motivation behind this change?


There are two test files marked by `#if 0` and `TODOs` when @RyanUnderhill changed CUDA code into a shared library https://github.com/microsoft/onnxruntime/pull/7626:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/providers/cuda/gemm_options_test.cc#L4-L5 && 
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/providers/cuda/reduction_functions_test.cc#L4-L5

The two test files are disabled at that time. These two files cannot be directly added back, because them do not fit into current build framework. This PR just moves them to the right place and refactor to reenable them.



@snnn 

> We do not have tests under core/providers/

There are really test cases under core/providers/, e.g. [core/providers/cuda/test/greedy_search_top_one.cc](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/test/greedy_search_top_one.cc). However, these tests are only enabled in `Debug` mode to reduce the Release lib size, which is introduced by https://github.com/microsoft/onnxruntime/pull/13016.

**Comment by pranavsharma** [comment]
> @snnn
> 
> > We do not have tests under core/providers/
> 
> There are really test cases under core/providers/, e.g. [core/providers/cuda/test/greedy_search_top_one.cc](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/test/greedy_search_top_one.cc). However, these tests are only enabled in `Debug` mode to reduce the Release lib size, which is introduced by #13016.

We don't really want tests to be located under these folders.

**Review by pranavsharma — CHANGES_REQUESTED** [review]
See my previous comment.

**Comment by guoyuhong** [comment]
@pranavsharma Thanks for the information. I was trying to run some CUDA unit-tests, but I found the cases in `test/providers/cuda/` is disabled because CUDA provider is separated to a dynamic lib and `onnxruntime_test_all` cannot directly call it. I found there are some UT cases in `core/providers/cuda/test/` but it is strange that it can run only in `Debug` mode. I'm trying to unify the UT for CUDA provider. 
You prefer the UT still in `test/providers/cuda/`, right? I still want to continue the work to refine the CUDA UTs.

**Comment by guoyuhong** [comment]
> Yes. Thanks! @RyanUnderhill please take a look at this as well. Tests that were _temporarily_ disabled as part of the cuda shared lib change have not been enabled again.

@pranavsharma Thanks for the clarification. This PR goes to the opposite side. I will take some time to change it back.

**Comment by guoyuhong** [comment]
@microsoft-github-policy-service terminate

**Comment by guoyuhong** [comment]
@microsoft-github-policy-service agree company="Ant Group"

**Comment by guoyuhong** [comment]
@pranavsharma @RyanUnderhill @snnn I have changed the PR to move all test cases from `core/providers/cuda/test/` to `test/providers/cuda/`. I verified that it is not possible to call the CUDA unittest from `onnxruntime_test_all` directly. That is why previously the CUDA unittest is implemented in `core/providers/cuda/test/` and to avoid binary size increasing, the test is only enabled in Debug mode. However, the previous implementation has draw backs.
1. If I want to run some tests, I need to change the mode to Debug, which is not convenient and need to rebuild the whole project.
2. The test cases cannot use google test macros, because the test is part of a cuda provider.

In this PR, I do following changes:
1. Add a new test lib `onnxruntime_providers_cuda_ut` which is similar to `onnxruntime_providers_cuda` but `onnxruntime_providers_cuda_ut` is only built if `onnxruntime_BUILD_UNIT_TESTS` is set. We can call all CUDA UTs through this ut lib without affecting production lib `onnxruntime_providers_cuda`.
2. Move all test cases from `core/providers/cuda/test/` to `test/providers/cuda/`. These test cases are built into lib `onnxruntime_providers_cuda_ut` and run by `./onnxruntime_test_all --gtest_filter="*CUDA_EP_Unittest*"`. Since the lib is only for test, we can use gtest macros in the test cases.
4. The cmake code in `cmake/onnxruntime_providers.cmake` is refactored a bit. A new function `onnxruntime_add_object_library` is to build a object target. The 2 libs `onnxruntime_providers_cuda_ut` & `onnxruntime_providers_cuda` share most of the code, so one object file can be used in both libs, which helps reduce build time. Another function `config_cuda_provider_shared_module` is used to configure all 3 similar targets.

Even through `onnxruntime_shared_lib_test` covers some test for CUDA provider. `onnxruntime_shared_lib_test` works like a coarse grain end-to-end test for CUDA provider. If CUDA unittest can run cases for a single component, this wound be helpful for CUDA developers.

Do you have suggestions for current code change?

**Comment by tianleiwu** [comment]
@guoyuhong, please merge the latest changes from main which includes a fix for training pipeline.

**Comment by guoyuhong** [comment]
@tianleiwu Thanks. I have rebased the latest main branch.

**Comment by tianleiwu** [comment]
@guoyuhong, you will need merge main again since the training pipeline was just fixed

**Comment by guoyuhong** [comment]
@tianleiwu Done.

**Comment by guoyuhong** [comment]
@tianleiwu There are 3 CI failures. I can only repo the two of `orttraining-linux-gpu-ci-pipeline`. `ONNX Runtime React Native CI Pipeline` failed in assembly language compiling. It seems not related to my change. I rebased the master to see whether the error is gone.

**Review by tianleiwu — APPROVED** [review]

**Comment by guoyuhong** [comment]
It looks like the CI failure does not relate to this PR. The latest main branch has the same failure.

**Comment by guoyuhong** [comment]
@tianleiwu I've rebased the main branch and resolve the conflict.

**Comment by tianleiwu** [comment]
@pranavsharma, please take a look since you requested changes.

**Review by pranavsharma — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/core/providers/cuda/cuda_provider_interface.cc [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include "core/session/onnxruntime_c_api.h"
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cuda/cuda_provider_interface.cc:1]: missed this: Please include Microsoft copyright notice and header here.

Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/cuda_provider_interface.cc:1]: @guoyuhong, please confirm it in https://github.com/microsoft/onnxruntime/pull/16437.

Comment (guoyuhong) [reply onnxruntime/core/providers/cuda/cuda_provider_interface.cc:1]: I will change this in later PR.

Comment (guoyuhong) [reply onnxruntime/core/providers/cuda/cuda_provider_interface.cc:1]: @tianleiwu has added the head in https://github.com/microsoft/onnxruntime/pull/16437 ;-)

**Comment by snnn** [comment]
No, it doesn't work. 

First, on Linux you duplicated the symbols. When we run onnxruntime_test_all.exe, we will have two copies of the CUDA EP in the same process. It could result hard to debug bugs. 
Second, in ORT we require have shared library has a version script file that will hide symbols by default.  It has multiple reasons. One of them is: if you don't do so, the shared library will have wired unloading behaviors: the library would be unloadable, and the destruction order of its global variables will mess up, which could lead to crashes on exit
Third, not only you duplicated the CUDA EP, on Linux you also duplicated gtest.  Again, nobody can explain what would happen. On Windows, if the gtest was built with abseil, it just always crashes.

**Pull request merged**